src/GraphQL/Types/ResolveFieldContextAdapter.cs

```diff
@@ -0,0 +1,58 @@
+using System.Collections.Generic;
+using System.Threading;
+using GraphQL.Instrumentation;
+using GraphQL.Language.AST;
+
+namespace GraphQL.Types
+{
+    internal class ResolveFieldContextAdapter<T> : IResolveFieldContext<T>
+    {
+        private readonly IResolveFieldContext _baseContext;
+
+        public ResolveFieldContextAdapter(IResolveFieldContext baseContext)
+        {
+            Source = (T)baseContext.Source; //will throw NullReferenceException or InvalidCastException if there's a problem
```

Comment:
OK, now I think that we should provide more specific error text to describe situation to user, not just NRE/InvalidCast. It needs to be done so that he understands what happened and what he needs to fix.

Reply:
Maybe `InvalidOperationException` or `ArgumentException`.

Reply:
After that final change I will be glad to merge this PR. Good work.

Reply:
Makes sense - I didn't think of the fact that the invalid cast exception could be triggered by 'user code'.  The constructor is a bit more wordy now but should produce an explanatory error message.